src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java

```diff
@@ -3775,7 +3775,7 @@ Symbol resolveSelf(DiagnosticPosition pos,
                 if (sym != null) {
                     if (staticOnly)
                         sym = new StaticError(sym);
-                    else if (env1.info.ctorPrologue)
+                    else if (env1.info.ctorPrologue && sym.kind == VAR && !isAllowedInPrologue(env1, (VarSymbol)sym))
```

Comment:
@lahodaj,

Thanks for your review.

> Can sym be something else then a variable here (given `name` can only be `this` or `super`)?

It's true that `name` can only be `this` or `super` - I've added an assertion to that effect in 887699b4085.

It's probably also true that `sym` has to be a `VarSymbol` but that was not immediately provable just from looking at the code. I've been burned before by making such assumptions (e.g., when things like `ResolveError` can sometimes pop up) so I'm playing it safe here.

Reply:
I may be too strict, but: I don't see a realistic way how it could be some kind of an error symbol here (as this fetches the symbol from a `Scope`). I may be wrong. But unless there's a known/realistic way how the symbol could be something else than a VarSymbol, I think I would prefer the code to fail one way or another here, rather than continuing. Because when the execution will continue with in an unpredicted situation, we don't know what will happen.

But, as I say, it may be too strict.

Reply:
OK, I've removed the (hopefully redundant) check in 619ab2b20be. I tried various erroneous inputs but could not trigger a compiler crash so it seems like you are right.